docarray/base_document/document.py

```diff
@@ -46,3 +47,165 @@ def __str__(self):
 
     def _get_string_for_regex_filter(self):
         return str(self)
+
+    def update(self, other: 'BaseDocument'):
+        """
+        Updates self with the content of other. Changes are applied to self.
+        Updating one Document with another consists in the following:
+         - setting data properties of the second Document to the first Document
+         if they are not None
+         - Concatenating lists and updating sets
+         - Updating recursively Documents and DocumentArrays
+         - Updating Dictionaries of the left with the right
+
+        It behaves as an update operation for Dictionaries, except that since
+        it is applied to a static schema type, the presence of the field is
+        given by the field not having a None value.
+
+            EXAMPLE USAGE
+
+            .. code-block:: python
+
+                from docarray import BaseDocument
+                from docarray.documents import Text
+
+
+                class MyDocument(BaseDocument):
+                    content: str
+                    title: Optional[str] = None
+                    tags_: List
+
+
+                doc1 = MyDocument(
+                    content='Core content of the document',
+                    title='Title',
+                    tags_=['python', 'AI']
+                )
+                doc2 = MyDocument(content='Core content updated', tags_=['docarray'])
+
+                doc1.update(doc2)
+                assert doc1.content == 'Core content updated'
+                assert doc1.title == 'Title'
+                assert doc1.tags_ == ['python', 'AI', 'docarray']
+
+        :param other: The Document with which to update the contents of this
+        """
+        if type(self) != type(other):
+            raise Exception(
+                f'Update operation can only be applied to '
+                f'Documents of the same type. '
+                f'Trying to update Document of type '
+                f'{type(self)} with Document of type '
+                f'{type(other)}'
+            )
+        from collections import namedtuple
+
+        from docarray import DocumentArray
+        from docarray.utils.reduce import reduce
+
+        # Declaring namedtuple()
+        _FieldGroups = namedtuple(
```

Comment:
We should declare this class outside of the the function IMO. So that we can use it as a type hint for the return type

I personnaly prefer the dataclass way of declaring the named tuple as you can declare the type hint

```python
class _FieldGroupds(NamedTuple):
    list_fields: List[str]
```

Reply:
this is a private method, so I do not see why it has to come outside